### PR TITLE
Closes #354 - Refactor google sign in

### LINKED ISSUE #354: Refactor Google sign in
Google sign in requires validator that would allow to register an account via Google that allows null value as password.

### PR DESCRIPTION
**Pull request by OleksandrMazur**

### CODE REVIEW

File: src/main/java/com/softserve/academy/spaced/repetition/utils/validators/PasswordNotNullValidator.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import com.softserve.academy.spaced.repetition.utils.validators.annotations.PasswordNotNull;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+public class PasswordNotNullValidator implements ConstraintValidator<PasswordNotNull, Account> {
+
+    @Override
+    public void initialize(PasswordNotNull constraintAnnotation) {
+        /**
```
Comment (antonbakalets) [inline src/main/java/com/softserve/academy/spaced/repetition/utils/validators/PasswordNotNullValidator.java]: Please move the JavaDoc out of method body. It should be writen before the method.

File: src/main/java/com/softserve/academy/spaced/repetition/utils/validators/PasswordNotNullValidator.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+         * Initializes the validator in preparation for isValid(Object, ConstraintValidatorContext) calls.
+         * This method is guaranteed to be called before any use of this instance for validation.
+         * @param constraintAnnotation
+         */
+    }
+
+    @Override
+    public boolean isValid(Account account, ConstraintValidatorContext constraintValidatorContext) {
+        boolean valid = false;
```
Comment (antonbakalets) [inline src/main/java/com/softserve/academy/spaced/repetition/utils/validators/PasswordNotNullValidator.java]: i like to write this kind of logical calculations in one line, like
```
return account.getAuthenticationType().equals(AuthenticationType.GOOGLE)
                || account.getAuthenticationType().equals(AuthenticationType.FACEBOOK)
                || (account.getPassword() != null;
```

You can try too.

File: src/main/java/com/softserve/academy/spaced/repetition/utils/validators/annotations/PasswordNotNull.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package com.softserve.academy.spaced.repetition.utils.validators.annotations;
+
+import com.softserve.academy.spaced.repetition.utils.validators.PasswordNotNullValidator;
+
+import javax.validation.Constraint;
+import javax.validation.Payload;
+import java.lang.annotation.*;
+
+@Target({ElementType.TYPE, ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER})
```
Comment (antonbakalets) [inline src/main/java/com/softserve/academy/spaced/repetition/utils/validators/annotations/PasswordNotNull.java]: I believe this annotation is expected only on class level.

File: src/main/java/com/softserve/academy/spaced/repetition/service/UserService.java [hunk 4]
```diff
@@ -166,6 +167,16 @@ void initializeNewUser(Account account, String email, AccountStatus accountStatu
...
 
+    /**
+     * Validates whether user's account can or cannot contain a null password value. If user registered
+     * via Facebook or Google then it allows for password field to be null, but will throw
+     * PasswordCannotBeNullException if user's authentication type is LOCAL and password is null.
+     *
+     * @param account            user`s account.
+     * @throws PasswordCannotBeNullException if authenticationType is LOCAL and password is null
+     */
+    void validateAccount(Account account) throws PasswordCannotBeNullException;
```
Comment (antonbakalets) [inline src/main/java/com/softserve/academy/spaced/repetition/service/UserService.java]: This method is called only inside the service implementation, therefore it should be removed from interface and marked private.
Please also move the JavaDoc.

**Review by antonbakalets — CHANGES_REQUESTED** [review]

**Review by antonbakalets — APPROVED** [review]

**Review by OlehPaliuh — APPROVED** [review]

**Review by OleksandrBartosh — APPROVED** [review]

**Pull request merged**